### PR TITLE
SciPy ndimage blog post

### PR DESCRIPTION
**Pull request by grlee77**
This is a blog post related to the SciPy ndimage work done in fall/winter 2020 under a NumFOCUS Small Development Grant.

attn: @rgommers

### CODE REVIEW

**Review by rgommers — COMMENTED** [review]
This looks great already @grlee77. The title and structure may need some thought - the title and start of the post read like a tutorial, and the switch to "now we're fixing bugs" is a bit abrupt.

Maybe just changing the title to "Making SciPy's image interpolation consistent" or something along those lines would help already.

File: posts/2021/01/scipy-ndimage-interpolation.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+At the top, the samples are treated as isolated points, whereas in the bottom panel points are considered as the centers of regulary spaced grid elements (i.e. pixels in 2D, voxels in 3D). The top interpolation has historically been used for ndimage's `wrap` (aka periodic) boundary mode. This leads to surprising behavior, because the period for `n` samples is `n - 1` with the first and last sample exactly overlapping. In SciPy 1.6, a new mode, `grid-wrap` was introduced which uses the grid interpretation of the lower panel so that all `n` samples are replicated.
+
+Similarly the `scipy.ndimage.zoom` function has a new `grid_mode` option that, when set to True, will cause resizing based on the full grid extent. With `grid_mode=True`, zoom's behavior becomes consistent with similar functions in scikit-image (`skimage.transform.resize`) and OpenCV (`cv2.resize`).
+
+### New documentation of boundary handling behavior
+
+Documentation of the boundary modes has been improved, adding [illustrations for the behavior of all modes](https://docs.scipy.org/doc/scipy/reference/tutorial/ndimage.html) and cross-linking to these figures from each individual function's docstring.
```
[the remark this answers is not in the document]
Comment (grlee77) [reply posts/2021/01/scipy-ndimage-interpolation.md:51]: yes, that was supposed to be the link. thanks

**Comment by grlee77** [comment]
I have pasted the scripts use to generate the images below. One of these used a synthetic dataset that is around 10MB in size, so I didn't want to commit that to the repo. I can add the scripts themselves somewhere if you like. 

Example image script

```Python
import scipy.ndimage as ndi
import matplotlib.pyplot as plt
from matplotlib import rcParams
import nibabel as nib

rcParams['image.interpolation'] = 'nearest'
rcParams['image.cmap'] = 'gray'

# t1_icbm_normal_1mm_pn1_rf0.mnc.gz synthetic data obtained from:
#    https://brainweb.bic.mni.mcgill.ca/selection_normal.html
#    (Modality=T1, Slice Thickness=1mm, Noise = 1%, non-uniformity=0%)
vol = nib.load('t1_icbm_normal_1mm_pn1_rf0.mnc.gz').get_fdata()

ds = 2
true = vol[:, ::-1, :][80, ...]
down = vol[::ds, ::-ds, ::ds][40, ...]
im0 = ndi.zoom(vol[::ds, ::-ds, ::ds], 2, order=0)[80, ...]
im1 = ndi.zoom(vol[::ds, ::-ds, ::ds], 2, order=1)[80, ...]
im3 = ndi.zoom(vol[::ds, ::-ds, ::ds], 2, order=3)[80, ...]

fig, axes = plt.subplots(1, 5)
for ax in axes:
    ax.set_axis_off()

vmin, vmax = true.min(), true.max()
axes[0].imshow(down, vmin=vmin, vmax=vmax)
axes[0].set_title('original\n(2 x 2 x 2 mm)')
axes[1].imshow(im0, vmin=vmin, vmax=vmax)
axes[1].set_title('upsampled\n(1 x 1 x 1 mm)\norder=0')
axes[2].imshow(im1, vmin=vmin, vmax=vmax)
axes[2].set_title('upsampled\n(1 x 1 x 1 mm)\norder=1')
axes[3].imshow(im3, vmin=vmin, vmax=vmax)
axes[3].set_title('upsampled\n(1 x 1 x 1 mm)\norder=3')
axes[4].imshow(true, vmin=vmin, vmax=vmax)
axes[4].set_title('ground truth\n\n(1 x 1 x 1 mm)')
fig.set_size_inches([12.04,  4.37])
plt.tight_layout()
plt.show()
```

# coordinate convention script
```Python
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rcParams

fig, axes = plt.subplots(2, 1, figsize=(5, 3), sharex=True)
n_cells = 7  # grid will be size (n_cells, n_cells)
n_cells_y = 1

# draw points
for ax in axes:
    ax.set_axis_off()
    # draw open circles at the locations of pixel centers
    for n in range(n_cells_y):
        ax.plot(np.arange(n_cells), -np.full(n_cells, n), 'ko')

# draw pixel borders
ax = axes[1]
for n in range(n_cells + 1):
    ax.plot([n - 0.5, n - 0.5], [0.5, -.5], 'k-')
ax.plot([-0.5, n_cells - .5], [-0.5, -0.5], 'k-')
ax.plot([-0.5, n_cells - .5], [0.5, 0.5], 'k-')

# draw arrows
w = .045
head_width = 3*w
head_length = 1.5*head_width
arrow_kwargs = dict(width=w, head_width=head_width, head_length=head_length,
                    shape='full', color='k')
axes[0].arrow(3, -0.4, 3 - head_length, 0, **arrow_kwargs)
axes[0].arrow(3, -0.4, -3 + head_length, 0, **arrow_kwargs)
text_kwargs = dict(fontdict=dict(fontsize=13, color='k'),
                   horizontalalignment='center')
axes[0].text(3, -0.85, 'period = n - 1', **text_kwargs)

axes[1].arrow(3, -0.75, 3.5 - head_length, 0, **arrow_kwargs)
axes[1].arrow(3, -0.75, -3.5 + head_length, 0, **arrow_kwargs)
axes[1].text(3, -1.2, 'period = n', **text_kwargs)

axes[0].set_xlim(axes[1].get_xlim())
axes[0].set_ylim(-1.25, 0.6)
axes[1].set_ylim(-1.25, 0.6)
title_dict = dict(fontsize=15, fontweight='bold')
axes[0].set_title('Point Model', fontdict=title_dict)

axes[1].set_title('Grid Model', fontdict=title_dict)

plt.tight_layout()
plt.show()
```

**Comment by rgommers** [comment]
> I have pasted the scripts use to generate the images below. One of these used a synthetic dataset that is around 10MB in size, so I didn't want to commit that to the repo. I can add the scripts themselves somewhere if you like.

Just in this PR comment sounds fine to me.

**Comment by rgommers** [comment]
Some more thoughts on the main messages of this post. You could emphasize:
- That the pixel/voxel vs. point data model wasn't well-defined, and that documenting this and making things consistent was an important contribution of your work.
- That `scipy.ndimage` hasn't had a dedicated maintainer for the past 15 year, plus that the C code is very hairy (if high-performant) which is why these fundamental issues never got addressed.

The "over a dozen issues" hints at the importance of the improvements, but doesn't make it explicit (there's lots of modules where we could close a dozen issues much more easily).

**Comment by grlee77** [comment]
Ralf, I am happy to add you as a co-author here if you like. Should we also try to get Stefan (or any other reviewers of the SciPy PRs) to review/contribute to the post? 

Also, I am thinking about a follow up post dealing with dask-image and CuPy as two downstream libraries implementing much of the scipy.ndimage API. If we write it in a more general / less technical focus it might fit under Quansight itself rather than Labs, but I am not sure yet.

**Comment by rgommers** [comment]
> Ralf, I am happy to add you as a co-author here if you like. 

Thanks, but I'd prefer for you to be the sole author - it's about your technical work, and you did all the writing too. All I did was throw in some review comments.

> Should we also try to get Stefan (or any other reviewers of the SciPy PRs) to review/contribute to the post?

If you want him to take a look please ask, but I don't think it's necessary.

> Also, I am thinking about a follow up post dealing with dask-image and CuPy as two downstream libraries implementing much of the scipy.ndimage API. If we write it in a more general / less technical focus it might fit under Quansight itself rather than Labs, but I am not sure yet.

That sounds great. I'd say if it's still talking about work you did it fits here, if it's more like "here's how you use this library to do X, Y or Z" then yes that is better on the Quansight blog.

**Comment by grlee77** [comment]
> If you want him to take a look please ask, but I don't think it's necessary.

Okay, let's just proceed without it then. I don't want to hold this up longer than necessary.

**Comment by rgommers** [comment]
In it goes. Thanks Greg!

**Pull request merged**